File: scottish_injuries_analysis.py
"""
Scottish Unintentional Injuries Analysis (Python)

Question 1: Among people aged 65+ in NHS Lothian and NHS Greater Glasgow and
Clyde, what were the four most common causes of unintentional-injury
hospital admissions in 2020/21, and how did the two health boards compare?

Data: Public Health Scotland "Unintentional Injuries" open dataset, joined
with the NHS Scotland health board reference table.
"""

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import json

# ---- Load and merge ------------------------------------------------------

injuries = pd.read_csv("data/scot_unintentional_injuries.csv")
healthboards = pd.read_csv("data/healthboards.csv")

merged_data = injuries.merge(healthboards, left_on="HBR", right_on="HB", how="left")
merged_data["NumberOfAdmissions"] = pd.to_numeric(merged_data["NumberOfAdmissions"], errors="coerce")
merged_data["FinancialYear"] = merged_data["FinancialYear"].astype("category")
merged_data["InjuryType"] = merged_data["InjuryType"].astype("category")
merged_data["AgeGroup"] = merged_data["AgeGroup"].astype("category")
merged_data["HBName"] = merged_data["HBName"].astype("category")

# ---- Filter to the population and boards of interest ---------------------

subset = merged_data[
    (merged_data["AgeGroup"].isin(["65-74 years", "75plus years"]))
    & (merged_data["FinancialYear"] == "2020/21")
    & (merged_data["HBName"].isin(["NHS Lothian", "NHS Greater Glasgow and Clyde"]))
].copy()

assert "NHS Lothian" in merged_data["HBName"].values
assert "NHS Greater Glasgow and Clyde" in merged_data["HBName"].values

# ---- Standardise inconsistent injury-type labels --------------------------

subset["InjuryType"] = (
    subset["InjuryType"]
    .str.strip()
    .str.lower()
    .str.title()
    .replace(
        {
            "Falling": "Falls",
            "Falling ": "Falls",
            "Poisoned": "Poisoning",
            "Poisoned ": "Poisoning",
            "Scalded": "Scalds",
            "Rta": "Collision-Related",
            "Struck By, Against": "Collision-Related",
            "Crushing": "Collision-Related",
            "Other ": "Other",
            "other": "Other",
            "other ": "Other",
        }
    )
)
subset["InjuryType"] = subset["InjuryType"].astype("category")

# ---- Group, rank, and plot the top 4 causes per board ---------------------

grouped = (
    subset.groupby(["HBName", "InjuryType"], observed=True)["NumberOfAdmissions"]
    .sum()
    .reset_index()
)

top4 = (
    grouped.sort_values("NumberOfAdmissions", ascending=False)
    .groupby("HBName")
    .head(4)
)

plt.figure(figsize=(8, 5))
sns.barplot(data=top4, x="NumberOfAdmissions", y="InjuryType", hue="HBName")
plt.title("Top 4 Causes of Injury Admissions (65+, 2020/21)")
plt.xlabel("Number of Admissions")
plt.ylabel("Injury Type")
plt.legend(title="Health Board")
plt.tight_layout()
plt.show()

# Finding: the two boards shared three of their top four causes
# (Falls, Accidental Exposure, Other), differing in the fourth:
# Collision-Related for NHS Lothian vs. Poisoning for NHS Greater
# Glasgow and Clyde.


# =====================================================================
# Question 2: Reconstructing a semi-structured JSON medication dataset
# Records are stored as unnamed lists of varying length because one
# field is missing (at an unpredictable position) in some rows. Rows
# are split into complete vs. short, the short ones are padded back to
# 10 fields at known positions, then both sets are combined into a
# typed DataFrame.
# =====================================================================


def load_json_file_named(file_name):
    """Load a JSON file from the data/ folder."""
    loaded_data = []
    file_location = f"data/{file_name}"
    try:
        with open(file_location, "r") as file:
            loaded_data = json.load(file)
    except OSError as e:
        print(f"Error. Does the file exist in this folder? {file_location}\n\n {e}")
    return loaded_data


medications = load_json_file_named("task2_data.json")

# Rows with all 10 fields (clean) vs. fewer fields (scrambled)
medications_clean = [row for row in medications if len(row) == 10]
medications_scrambled = [row for row in medications if len(row) != 10]

# Missing-field position for each scrambled row (1-based index), found by
# inspecting medications_scrambled above
insertion_positions = [5, 8, 5, 6, 8, 5, 5, 5]

corrected_rows = []
for row, pos in zip(medications_scrambled, insertion_positions):
    padded = row[: pos - 1] + [None] + row[pos - 1 :]
    corrected_rows.append(padded)

col_names = [
    "BNF_Code", "PrescriptionDate", "Source", "Dose", "Unit",
    "Colour", "Shape", "Age", "FridgeFlag", "Frequency",
]

all_rows = medications_clean + corrected_rows
df = pd.DataFrame(all_rows, columns=col_names)

df_cleaned = df.copy().assign(
    PrescriptionDate=pd.to_datetime(df["PrescriptionDate"], format="%d%m%Y", errors="coerce"),
    Dose=pd.to_numeric(df["Dose"], errors="coerce"),
)

df_final = df_cleaned.drop(columns=["Age", "Shape", "Colour"])
df_final = df_final.sort_values(by="PrescriptionDate")
df_final[["Source", "Unit"]] = df_final[["Source", "Unit"]].astype(
    {"Source": "category", "Unit": "category"}
)

print(df_final.info())

File: scottish_injuries_analysis.R
# Scottish Unintentional Injuries Analysis (R)
#
# Question 1: Across Scotland, what are the most and least common
# unintentional injuries for young people aged 5-14 in 2013/14 and 2022/23,
# and did admission rates change between the two periods?
#
# Data: Public Health Scotland "Unintentional Injuries" open dataset,
# joined with the NHS Scotland health board reference table.

library(tidyverse)
library(lubridate)
library(rjson)
library(ggplot2)
library(forcats)

# ---- Load data ---------------------------------------------------------

injuries <- read_csv("data/scot_unintentional_injuries.csv")
healthboards <- read_csv("data/healthboards.csv")

# ---- Join and type-cast -------------------------------------------------

joined_df <- left_join(injuries, healthboards, by = c("HBR" = "HB"))

correct_type <- joined_df %>%
  mutate(
    FinancialYear = as.factor(FinancialYear),
    AgeGroup = as.factor(AgeGroup),
    InjuryType = as.factor(InjuryType),
    NumberOfAdmissions = as.integer(NumberOfAdmissions),
    HBR = as.factor(HBR),
    CountryName = as.factor(CountryName)
  )

# ---- Filter to the population and years of interest ---------------------

minimised_df <- correct_type %>%
  select(FinancialYear, HBR, AgeGroup, InjuryType, NumberOfAdmissions, CountryName) %>%
  filter(
    CountryName == "Scotland",
    AgeGroup %in% c("5-9 years", "10-14 years"),
    FinancialYear %in% c("2013/14", "2022/23")
  )

# ---- Standardise inconsistent injury-type labels -------------------------

cleaned <- minimised_df %>%
  mutate(InjuryType = fct_recode(InjuryType,
    "Falls" = "Falling",
    "Poisoning" = "Poisoned",
    "Scalds" = "Scalded",
    "Collision-related" = "Struck by, against",
    "Collision-related" = "RTA",
    "Collision-related" = "Crushing"
  ))

# ---- Summarise and plot --------------------------------------------------

cleaned %>%
  group_by(FinancialYear, InjuryType) %>%
  summarise(Admissions = sum(NumberOfAdmissions, na.rm = TRUE), .groups = "drop") %>%
  ggplot(aes(x = reorder(InjuryType, Admissions), y = Admissions, fill = FinancialYear)) +
  geom_col(position = "dodge") +
  coord_flip() +
  labs(
    title = "Injury Admissions for Ages 5-14 in 2013/14 and 2022/23",
    x = "Injury Type",
    y = "Number of Admissions"
  ) +
  theme_minimal()

# Finding: "Falls" was the most common cause of admission in both years;
# "Scalds" and "Poisoning" were the least common. Falls admissions fell
# slightly over the period while some other categories increased.


# =====================================================================
# Question 2: Reconstructing a semi-structured JSON medication dataset
# Some records are missing a field at an unpredictable position, which
# misaligns a naive row-bind. Rows are split into complete vs. short,
# the short ones are padded back to 10 fields at known positions, then
# both sets are recombined and given correct column types.
# =====================================================================

medications <- fromJSON(file = "data/task2_data.json")

# Split into complete (10-field) rows and short/scrambled rows
medications_clean_list <- medications[lengths(medications) == 10]
medications_scrambled_list <- medications[lengths(medications) != 10]

medications_clean_tibble <- as_tibble(
  do.call(rbind, medications_clean_list),
  .name_repair = "minimal"
)
colnames(medications_clean_tibble) <- c(
  "BNF_Code", "PrescriptionDate", "Source", "Dose", "Unit",
  "Colour", "Shape", "Age", "FridgeFlag", "Frequency"
)

# Inspect the scrambled rows to find which field index is missing in each
medications_scrambled_list[]

# Known missing-field position for each scrambled row (1-based), found by
# inspection above. Pad each row with NA at that position.
positions <- c(5, 8, 5, 6, 8, 5, 5, 5)
values_to_insert <- rep(list(NA), length(positions))

for (i in seq_along(medications_scrambled_list)) {
  medications_scrambled_list[[i]] <- medications_scrambled_list[[i]] %>%
    append(values_to_insert[[i]], after = positions[i] - 1)
}

medications_corrected_tibble <- as_tibble(
  do.call(rbind, medications_scrambled_list),
  .name_repair = "minimal"
)
colnames(medications_corrected_tibble) <- c(
  "BNF_Code", "PrescriptionDate", "Source", "Dose", "Unit",
  "Colour", "Shape", "Age", "FridgeFlag", "Frequency"
)

# Combine, drop columns not needed downstream, and assign correct types
df_selected <- bind_rows(medications_clean_tibble, medications_corrected_tibble) %>%
  select(-Colour, -Shape, -Age) %>%
  mutate(
    PrescriptionDate = dmy(PrescriptionDate),
    Dose = as.numeric(Dose),
    Unit = as.factor(Unit),
    Frequency = as.integer(str_remove(Frequency, "x")),
    Source = str_to_lower(Source),
    Source = case_when(
      Source %in% c("hosp", "hopsital", "hospital") ~ "hospital",
      Source %in% c("pharm", "pharmacy") ~ "pharmacy",
      Source %in% c("dent", "dentist") ~ "dentist",
      TRUE ~ Source
    ),
    Source = factor(Source),
    FridgeFlag = as.logical(as.integer(FridgeFlag))
  )

glimpse(df_selected)
